pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,1566 @@
+        /// Commission due to validators, set at genesis
+        type DefaultValidatorCommission: Get<Perbill>;
+        /// Minimum stake required for any account to be in `SelectedCandidates` for the round
+        type MinValidatorStake: Get<BalanceOf<Self>>;
+        /// Minimum stake required for any account to be a validator candidate
+        type MinValidatorPoolStake: Get<BalanceOf<Self>>;
+        /// Minimum stake for any registered on-chain account to nominate
+        type MinNomination: Get<BalanceOf<Self>>;
+        /// Minimum stake for any registered on-chain account to become a nominator
+        type MinNominatorStake: Get<BalanceOf<Self>>;
+		/// Tokens have been minted and are unused for validator-reward.
+		/// See [Era payout](./index.html#era-payout).
+		type RewardRemainder: OnUnbalanced<NegativeImbalanceOf<Self>>;
+		/// This pallet's module id. Used to derivate a dedicated account id to store session
+		/// rewards for validators and nominators in.
+		type PalletId: Get<ModuleId>;
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_finalize(n: T::BlockNumber) {
+
+            // let mut round = <Round<T>>::get();
+            // if round.should_update(n) {
+            //     // mutate round
+            //     round.update(n);
+            //     // pay all stakers for T::BondDuration rounds ago
+            //     Self::pay_stakers(round.current);
+            //     // start next round
+            //     <Round<T>>::put(round);
+            //     // snapshot total stake
+            //     <Staked<T>>::insert(round.current, <Total<T>>::get());
+            //     // Self::deposit_event(Event::NewRound(
+            //     // 	round.first,
+            //     // 	round.current,
+            //     // 	collator_count,
+            //     // 	total_staked,
+            //     // ));
+            // }
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Set the validators who cannot be slashed (if any).
+        ///
+        /// The dispatch origin must be Root.
+        #[pallet::weight(10_000)]
+        pub fn set_invulnerables(
+            origin: OriginFor<T>,
+            invulnerables: Vec<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+            <Invulnerables<T>>::put(&invulnerables);
+            Self::deposit_event(Event::NewInvulnerables(invulnerables));
+            Ok(().into())
+        }
+        #[pallet::weight(10_000)]
+        /// Set the total number of validator candidates selected per round
+        /// - changes are not applied until the start of the next round
+        pub fn set_total_validator_per_round(
+            origin: OriginFor<T>,
+            new: u32,
+        ) -> DispatchResultWithPostInfo {
+            frame_system::ensure_root(origin)?;
```

Comment:
same regarding origins

Reply:
:+1:  Done

Origin of the call may be either Root or CancelOrigin

```
            T::CancelOrigin::try_origin(origin)
                .map(|_| ())
                .or_else(ensure_root)?;

```